README
Mail::SimpleList README
-----------------------

History:

	Version 0.88, Fri Mar 26 04:29:29 UTC 2004, with two nice bugfixes
	Version 0.87, Mon Mar 15 04:54:04 UTC 2004
	Version 0.86, Tue Dec 30 00:21:24 UTC 2003
	Version 0.81, Tue Oct 14 23:09:53 UTC 2003
	Version 0.80, Sun Sep 21 20:41:18 UTC 2003
	Version 0.79, Fri Jul  4 19:14:54 GMT 2003
	Version 0.78, Mon Jun 30 04:51:58 GMT 2003
	Version 0.75, Sat Jun 28 20:13:22 GMT 2003, back home
	Version 0.70, Sun Jun 15 19:11:31 GMT 2003, somewhere over Alabama (maybe)

Mail::SimpleList is a set of modules for creating and using small, simple
mailing lists.  They're intended to be temporary and very easy to use.  They're
not tremendously featureful, but they work pretty well.

The idea came about one afternoon when we were debating what to do after work
on Friday.  We'd sent several mails back and forth and the Cc line had grown
unwieldy.  One person said, "Take me off the list; I'll be out of town", but
not everyone had received that message yet and the replies were getting
unwieldy.

What we needed was a simple way to start a small mailing list that would only
be active for a few days, where everyone could subscribe or unsubscribe at
will.

That's what Mail::SimpleList does.

For more information, feel free to consult the talk I gave to the Portland Perl Mongers (http://pdx.pm.org/) on 11 June 2003.  The slides are available at:

	http://wgz.org/chromatic/talks/MailSimpleList/

INSTALLATION
------------

First, run the Build.PL file to create the Build program.  Then install as
normal:

	$ ./Build
	$ ./Build test
	$ sudo ./Build install

All of the tests should pass.  If not, please send the complete diagnostic
output to the author and we'll try to fix the bugs.  This should work on any
Unix system and probably elsewhere, but that'll depend on several other
components.

To use this program, you need to create a new account on your mail server.  I
use 'alias'.  This user also needs a writeable directory.  I use
/home/alias/aliases/.  Make sure this directory exists and is writeable by this
user.

This user must also be able to receive e-mail and execute a procmail-style
.forward or .qmail file.

This user must also receive mail with an optional delimiter.  For example, the
user 'alias' must be able to receive mail to 'alias+somelistsname'.  The
extended name must be passed through to the program for mail to reach any
lists.

See the examplefilter.pl file for the recommended way to execute the mailing
list manager.  The $alias_dir variable must contain the absolute path to the
aliases directory -- if this path is incomplete, the program will fail.  Put
this file in the alias user's path and make sure it is executable.  I like
/home/alias/bin/simplelist.pl.

Your .forward or .qmail file should contain something like the following:

	# .forward style
	| "/home/alias/bin/simplelist.pl 2>> simplelist.log"

or

	# .qmail style
	preline /home/alias/bin/simplelist.pl 2>> simplelist.log

Please consult the documentation for your mail delivery program if neither of
these work.  The log file can be handy for determining path and permissions
problems.

USAGE
-----

This information is also contained within the POD of Mail::SimpleList:

	$ perldoc Mail::SimpleList

Creating a list is simple.  Just send an e-mail to the alias user on your host.
To create a list, the mail subject must start with *new*.  The body of the list
should contain a list of e-mail addresses to subscribe to the list.  As the
sender, you will be marked as the list owner and you will be subscribed
automatically.

The list creator will receive an e-mail if the list is created successfully.
It will include the posting address of the list.  All other users will be
notified that they have been subscribed to the list.

Posting to the list is simple; just send a message to the list alias.  Anyone
included in the To: or Cc: lines of the e-mail will automatically be subscribed
to the list.

Unsubscribing from a list is also simple.  Send a message to the list alias
with a subject starting with *UNSUBSCRIBE*.  You should receive a success
message.


DIRECTIVES
----------

As of version 0.60, you can also specify additional directives in the body of
the creation message.  The useful directives right now are 'Expires',
'Auto_add', 'Closed', and 'Description'.  Directives precede any subscribable
addresses.  They take the form:

	Directive: argument string

Aliases can be closed to non-members.  This directive:

	Closed: yes

will create a list that does not allow messages to be posted from
non-subscribers.  They can receive mail also sent to the alias, however.

You can create an alias that will accept messages for a limited period of time.
The Expires directive takes a time string when to expire the alias in the form
of Months, weeks, days, hours, and minutes.  Thus this directive:

	Expires: 1M2w3d4h

will make the alias expire in 1 Month, 2 weeks, 3 days, and 4 hours.  For the
purpose of simplicity, one Month is equal to thirty days.

Messages sent to the alias after the expiration period is over will not be
delivered.  The sender will receive a message to this effect.

Aliases automatically subscribe addresses found in the Cc header by default.
The Auto_add directive controls this.  Saying:

	Auto_add: no

will disable this feature.

Added addresses will receive the standard subscription notice and they will be
removed from the headers of the message, so as not to receive duplicate
messages.  For now, these addresses must appear in the Cc header, not the To
header.

Description is a simple, one-line description of the purpose of the alias.  It
will be sent to everyone who is added to the alias.

	Description: Let's have a cupcake party on 8 September.

Name is an optional name for the alias.  If specified, it will be used instead
of the randomly generated alias.

	Name: jacobs-birthday

ADVANCED USAGE
--------------

The auto-generated list aliases can be unwieldy.  Since everything is stored in
simple text files, one per list, you can create a symlink to an alias with a
nicer name and post to that list instead.  This is much nicer for
longer-lasting lists.

If subscribing and unsubscribing is too difficult, you can edit the files
manually, if you know YAML sufficiently well.  It's really easy, so it
shouldn't be much trouble.

ENHANCEMENTS
-------------

This is intended to be a simple program with a simple purpose.  It does just
about everything I could ever want right now, but there are a few improvements
up my sleeve.

Any address in the To header besides the alias address appears to wreak some
havoc with the program.  Try to avoid this for now.

HELP IS HERE as of version 0.80.

There's not much security on lists; address spoofing is possible.  I'd like to
add optional cryptographic signing.

There's no archiving yet, but it should be fairly easy to add.

There are a few hard-coded things, such as the alias delimiter.  That should be configurable.

There might be room for a few more tests.

LICENSE, COPYRIGHT, and CREDITS
-------------------------------

Copyright 2003 - 2004, chromatic <chromatic@wgz.org>

This software is made available under the same license as Perl.

This software would not have been possible without the inspiration of Simon
Cozens' Mail::Audit.  The Portland Perl Mongers specified many improvements.
Allison Randal had a neat idea for giving aliases better names.


Changes
0.95
    Sun Jul 28 2013
    - fixed documentation link (Mike Flannagin)

0.94
    Tue Jul 05 19:17:00 PDT 2013
    - fixed tests for Perl 5.24 and, more importantly, Email::MIME
    - added to Github

0.93
    Mon Nov 27 06:05:35 UTC 2006 ($Rev: 60 $, $Author: chromatic $)
    - fix a test that failed with a newer Test::MockObject
    - make all tests run under taint mode
    - add POD tests to run under developer mode

0.92
    Fri Nov 26 20:35:33 UTC 2004 (Rev: 55, Author: chromatic)
    - update Build.PL requirement on Mail::Action 0.42

    Thu Nov 25 20:21:53 UTC 2004 (Rev: 52, Author: chromatic)
    - respect the Name: directive when cloning an alias (Brett Warden)
    - more M::A::R changes (consistify Email::Address use)
    - improve FakeMail for M::A::R changes
    - added expansion_pattern() for overriders (Andy Lester)
    - bumped up submodule version numbers

0.91
    Tue Sep 28 23:29:45 UTC 2004 (Rev: 47, Author: chromatic)
    - change Mail::SimpleList to use Mail::Action::Request
    - be more aggressive to find correct delivery alias
    - simplify tests

0.90
    Sun Jul 11 03:32:51 UTC 2004 (Rev: 40, Author: chromatic)
    - added raw_message() to t/lib/FakeMail.pm
    - ported to Mail::Action 0.40 (Email::Simple, Email::MIME)
    - added Mail::SimpleList::remove_alias_from()
    - marked build requirements in Build.PL separately

0.88
    Fri Mar 26 04:28:34 UTC 2004 (Rev: 37, Author: chromatic)
    - update version number for release

    Fri Mar 26 04:26:31 UTC 2004 (Rev: 36, Author: chromatic)
    - delete Delivered-To header in Mail::SimpleList::deliver() (bounce bug fix)
    - delineate unsubscribe message better

0.87
    Mon Mar 15 20:27:25 UTC 2004 (Rev: 33, Author: chromatic)
    - fix a documentation typo (Brett Warden)

    Mon Mar 15 06:10:52 UTC 2004 (Rev: 32, Author: chromatic)
    - mark dependency on Mail::Action 0.30 in Build.PL

    Mon Mar 15 05:57:26 UTC 2004 (Rev: 31, Author: chromatic)
    - fix spelling of Simon's name in README (sorry)

    Mon Mar 15 02:35:51 UTC 2004 (Rev: 31, Author: chromatic)
    - improve multi-part message handling, especially with sigs
    - bump version number
    - update examplefilter.pl to show procmail workaround

    Thu Mar 11 01:23:35 UTC 2004 (Rev: 31, Author: chromatic)
    - ported to new Mail::Action (using Mail::Message, not Mail::Internet)
    - set author keyword expansion on Changes

0.86 
    Tue Dec 30 00:21:35 UTC 2003 (Rev: 27)
    - fix portability of t/aliases.t
    - should pass tests on non-Unix platforms now

0.85
    Sun Dec 14 01:00:13 UTC 2003 (Rev: 26)
    - update dependency on Mail::Action

    Sat Dec 13 22:37:22 UTC 2003 (Rev: 25)
    - use Mail::Action::Address roles in Mail::SimpleList::Alias
    - add dependency on Class::Roles

    Thu Dec 11 05:40:30 UTC 2003 (Rev: 24)
    - factored out Mail::Action::Address
    - update files for 0.85 release

0.82
    Sun Nov  2 04:06:50 UTC 2003 (Rev: 23)
    - factored out Mail::Action::Storage
    - made Mail::SimpleList::Aliases extend Mail::Action::Storage
    - fixed typo in previous Changes version

    Sun Nov  2 03:29:02 UTC 2003 (Rev: 22)
    - factored out Mail::Action
    - made Mail::SimpleList extend Mail::Action

0.81 Tue Oct 14 23:05:08 UTC 2003
    - ignore signatures when creating new lists (Brett Warden)
    - update files for 0.81 release

0.80 Sun Sep 21 20:42:34 UTC 2003
    - added lib/Mail/SimpleList/PodToHelp and t/podtohelp.t
    - fill in command_help() in Mail::SimpleList
    - update files for 0.80 release

     Sun Sep 21 07:29:01 UTC 2003
    - set Reply-To list address for normal responses
    - fix caplitalization of CC header (docs and tests)

0.79 Sun Sep 21 07:07:35 UTC 2003
    - fix POD error in Mail::SimpleList::Alias

     Fri Jul  4 19:14:54 GMT 2003
    - be sure to save list after auto_add
    - remove Cc: headers appropriately after auto_add

     Thu Jul  3 06:18:04 GMT 2003
    - add List-Id and X-MSL-Seen headers
    - disallow sending the same message to the alias twice
    - allow the To header to remain
    - allow the Date header to remain (Tim Maher and Allison Randal suggestion)
    - disallow the 'From ' mbox heading (Tim Maher helped debug this one)

0.78 Mon Jun 30 04:54:27 GMT 2003
    - added Name directive to specify alias name (Allison Randal suggestion)
    - updated docs to reflect this
    - minor code shakeup to make it possible
        - Mail::SimpleList::generate_alias()
        - Mail::SimpleList::post_address()
        - Mail::SimpleList::process_body()
        - Mail::SimpleList::Aliases::create() -- no longer saves

0.75 Sat Jun 28 20:12:54 GMT 2003
    - added Description directive and documentation
    - removed unused code from t/aliases.t
    - copy all headers except the ones that need changing

0.71 Sat Sep 20 23:47:10 UTC 2003
    - fix hash order dependent test #57 in t/base.t
    - fix t/lib/FakeIn.pm to generate better glob refs

0.70 Sun Jun 15 19:14:08 GMT 2003
    - added documentation
    - added link to Portland Perl Mongers talk in README
    - added support for list cloning
    - miscellaneous improvements to acceptance/customer tests

0.65 4 June 2003 (or thereabouts)
    - improved tests with t/lib/FakeMail
    - clean up after tests (remove t/alias)
    - ensure that generated aliases never overlap (as much as possible)
    - add alias cloning

0.60 Fri May 30 01:35:20 GMT 2003
    - added option for closed aliases
    - added option for no auto add

     Thu May 29 06:56:08 GMT 2003
    - added support for alias expiration
    - fixed the bug that ate message bodies

0.50 Mon May 26 01:36:43 GMT 2003
    - initial release
    - handles creating, replying, and unsubscribing

pre-release
    - was called something else
    - wasn't quite as nice


Build.PL
#! perl

use Module::Build;

my $class = Module::Build->subclass(
	class => 'Module::Build::FilterTests',
	code  => <<'END_HERE',
	
	use File::Glob;
	use File::Spec::Functions;
	
	sub ACTION_disttest
	{
		my $self = shift;
		local $ENV{PERL_RUN_ALL_TESTS} = 1;
		$self->SUPER::ACTION_disttest( @_ );
	}
	
	sub find_test_files
	{
		my $self  = shift;
		my $tests = $self->SUPER::find_test_files( @_ );
	    
		return $tests unless $ENV{PERL_RUN_ALL_TESTS};
	
		my $test_pattern = catfile(qw( t developer *.t ) );
		unshift @$tests, File::Glob::bsd_glob( $test_pattern );
		return $tests;
	}
END_HERE
);

my $build = $class->new(
	module_name => 'Mail::SimpleList',
	license     => 'perl',
	requires    =>
	{
		Class::Roles     => '0.30',
		Mail::Action     => '0.42',
		Mail::Internet   => '1.53',
		Pod::Simple      => '',
		YAML             => '0.35',
	},
	build_requires       =>
	{
		Test::Exception  => '',
		Test::MockObject => '1.07',
		Test::Simple     => '0.47',
	},
	dist_version_from => 'lib/Mail/SimpleList.pm',
	dist_author       => 'chromatic <chromatic@wgz.org>',
);

$build->create_build_script();


lib/Mail/SimpleList/Aliases.pm
package Mail::SimpleList::Aliases;

use strict;
use base 'Mail::Action::Storage';

use File::Spec;

use Mail::SimpleList::Alias;

use vars qw( $VERSION );
$VERSION = '0.95';

sub new
{
    my ($class, $directory) = @_;
    $directory ||= File::Spec->catdir( $ENV{HOME}, '.aliases' );

    $class->SUPER::new( $directory );
}

sub stored_class
{
    'Mail::SimpleList::Alias';
}

sub storage_extension
{
    'sml'
}

sub create
{
    my ($self, $owner) = @_;

    return Mail::SimpleList::Alias->new(
        owner   => $owner,
        members => [$owner],
    );
}

1;

__END__

=head1 NAME

Mail::SimpleList::Aliases - manages Mail::SimpleList::Alias objects

=head1 SYNOPSIS

    use Mail::SimpleList::Aliases;
    my $aliases = Mail::SimpleList::Aliases->new( '.aliases' );

=head1 DESCRIPTION

Mail::SimpleList::Aliases manages the creation, loading, and saving of
Mail::SimpleList::Alias objects.  If you'd like to change how these objects are
managed on your system, subclass or reimplement this module.

=head1 METHODS

=over 4

=item * new( [ $alias_directory ] )

Creates a new Mail::SimpleList::Aliases object.  The single argument is
optional but highly recommended.  It should be the path to where Alias data
files are stored.  Beware that in filter mode, relative paths can be terribly
ambiguous.

If no argument is provided, this will default to C<~/.aliases> for the invoking
user.

=item * storage_dir()

Returns the directory where this object's Alias data files are stored.

=item * exists( $alias_name )

Returns true or false if an alias with this name exists.

=item * fetch( $alias_name )

Creates and returns a Mail::SimpleList::Alias object represented by this alias
name.  This can return nothing if the alias does not exist.

=item * create( $owner )

Creates and returns a new Mail::SimpleList::Alias object, setting the owner.
Note that you will need to C<save()> the object yourself, if that's important
to you.

=item * save( $alias, $alias_name )

Saves a Mail::SimpleList::Alias object provided as C<$alias> with the given
name in C<$alias_name>.

=back

=head1 AUTHOR

chromatic, C<chromatic@wgz.org>, with helpful suggestions from friends, family,
and peers.

=head1 BUGS

None known.

=head1 TODO

No plans.  It's pretty nice as it is.

=head1 SEE ALSO

L<Mail::Action::Storage>, the parent class.

=head1 COPYRIGHT

Copyright (c) 2016, chromatic.  All rights reserved.  This module is
distributed under the same terms as Perl itself.  Convenient for you!


lib/Mail/SimpleList.pm
package Mail::SimpleList;

use strict;
my $pod = do { local $/; <DATA> };

use base 'Mail::Action';
use Carp 'croak';

use Mail::Mailer;
use Email::Address;
use Email::MIME;

use vars '$VERSION';
$VERSION = '0.95';

use Mail::SimpleList::Aliases;

sub storage_class
{
    'Mail::SimpleList::Aliases'
}

sub parse_alias
{
    my ($self, $address)  = @_;
    my ($add)             = Email::Address->parse( $address );
    my $user              = $add->user();
    my $expansion_pattern = $self->expansion_pattern();

    return ( $user =~ $expansion_pattern ) ? $1 : '';
}

sub expansion_pattern
{
    return qr/\+([^+]+)$/;
}

sub command_help
{
    my $self   = shift;
    $self->SUPER::command_help( $pod, 'USING LISTS', 'DIRECTIVES' );
}

sub command_new
{
    my $self      = shift;
    my $from      = $self->address_field( 'From' );
    my $addresses = $self->storage();
    my $alias     = $addresses->create( $from );
    my $users     = $self->process_body( $alias );
    my $id        = $self->generate_alias( $alias->name() );
    my $post      = $self->post_address( $id );

    $self->add_to_alias( $alias, $post, @$users );
    $addresses->save( $alias, $id );

    $self->reply({ To => $from, Subject => "Created list $id" },
        "Mailing list created.  Post to $post." );

    return $alias;
}

sub command_clone
{
    my $self       = shift;

    my $from       = $self->address_field( 'From' );
    my $request    = $self->request();
    (my $subject   = $request->header( 'Subject' )) =~ s/^\*clone\*\s+//;
    my ($alias_id) = $self->parse_alias( $subject );
    my $addresses  = $self->storage();
    my $parent     = $addresses->fetch( $alias_id );
    my $alias      = $addresses->create( $from );
    my $users      = $self->process_body( $alias );
    my $wanted_id  = $alias->name() || $alias_id;
    my $id         = $self->generate_alias( $wanted_id );
    my $post       = $self->post_address( $id );

    $self->add_to_alias( $alias, $post, @{ $parent->members() }, @$users );

    $addresses->save( $alias, $id );

    $self->reply({ To => $from, Subject => "Cloned alias $alias_id => $id" },
        "Mailing list created.  Post to $post." );

    return $alias;
}

sub address_field
{
    my ($self, $field) = @_;

    my @values         = $self->request->header( $field   );
    return wantarray ? @values : $values[0]->address();
}

sub generate_alias
{
    my ($self, $id) = @_;
    my $addresses   = $self->storage();

    $id      ||= sprintf '%x', reverse scalar time;

    while ($addresses->exists( $id ))
    {
        $id    = sprintf '%x', ( reverse ( time() + rand($$) ));
    }

    return $id;
}

sub post_address
{
    my ($self, $id)   = @_;
    my ($address)     = $self->address_field( 'To' );

    # if this is a *new* request, there's no To field anymore
    $address        ||= $self->request->recipient();
    my $host          = $address->host();
    (my $base         = $address->user()) =~ s/\+([^+]+)$//;

    return "$base+$id\@$host";
}

sub reply
{
    my ($self, $headers) = splice( @_, 0, 2 );
    $headers->{'X-MSL-Seen'}    = '1';
    $self->SUPER::reply( $headers, @_ );
}

sub command_unsubscribe
{
    my $self         = shift;
    my ($alias, $id) = $self->fetch_address();
    my $from         = $self->request->header( 'From' )->address();

    $self->reply({ To => $from, Subject => "Remove from $alias" },
         ($alias->remove_address( $from ) and
          $self->storage->save( $alias, $id )) ?
            "Unsubscribed $from successfully." :
            "Unsubscribe unsuccessful for $from.  Check the address."
    );
}

sub process
{
    my $self    = shift;

    return if $self->request->header('X-MSL-Seen');
    my $command = $self->find_command();
    return $self->$command() if $command;

    my $alias   = $self->fetch_address();
    return $self->deliver( $alias ) if $alias;
    $self->reject();
}

sub deliver
{
    my ($self, $alias) = @_;

    my $name       = $alias->name();
    my $request    = $self->request();
    my $recipient  = $request->recipient();
    my $sent_to    = $recipient->address();
    my $host       = $recipient->host();
    my $message    = $request->copy_headers();
    $message->{To} = $sent_to;

    unless ($self->can_deliver( $alias, $message ))
    {
        my $body       = delete $message->{Body};
        $message->{To} = delete $message->{From};
        $self->reply( $message, $body );
        return;
    }

    my $desc    = $alias->description() || '';

    if ( $alias->auto_add() )
    {
        my @to_friends = map { $_->address() } $request->header( 'To' );
        my @cc_friends = map { $_->address() } $request->header( 'Cc' );

        $self->add_to_alias( $alias, @to_friends, @cc_friends );
        $self->storage->save( $alias, $name );
    }

    $message->{Bcc}        = $alias->members();
    $message->{'List-Id'}  = ( $desc ? qq|"$desc" | : '') .
        "<$name.list-id.$host>"; 
    $message->{'Reply-To'} = $sent_to;
    delete $message->{'Delivered-to'};

    my $body    = $self->add_signature( "\n-- \nTo unsubscribe:" .
        qq| reply to this sender alone with "*UNSUBSCRIBE*" in the subject.\n|
    );

    $self->reply( $message, $body );
}

sub add_signature
{
    my ($self, $sig)  = @_;
    my $request       = $self->request();
    my @parts         = $request->message->parts();

    if (@parts == 1)
    {
        $request->message->body_set( $request->message->body() . $sig );
    }
    else
    {
        my $sig_part  = Email::MIME->create(
            attributes => {
                encoding     => '7bit',
                disposition  => 'attachment',
                content_type => 'text/plain',
            },
            body => $sig,
        );

        push @parts, $sig_part;
        $request->message->parts_set( \@parts );
    }

    return $request->message->body_raw();
}

sub reject
{
    my $reason = $_[1] || "Invalid alias\n";
    $! = 100;
    die $reason;
}

sub notify
{
    my ($self, $alias, $id) = splice( @_, 0, 3 );

    my $owner = $alias->owner();
    my $desc  = $alias->description();

    for my $address ( @_ )
    {
        $self->reply({
            From       => $owner,
            To         => $address, 
            'Reply-To' => $id,
            Subject    => "Added to alias $id",
        }, "You have been subscribed to alias $id by $owner.\n\n", $desc );
    }
}

sub can_deliver
{
    my ($self, $alias, $message) = @_;
    if ( $alias->closed() and not
        grep { $_ eq $message->{From} } @{ $alias->members() })
    {
        $message->{To}      = $message->{From};
        $message->{Subject} = 'Alias closed';
        $message->{Body}    = 'This alias is closed to non-members.';
        return;
    }
    return 1 unless my $expires = $alias->expires();
    if ($expires < time())
    {
        $message->{To}      = $message->{From};
        $message->{Subject} = 'Alias expired';
        $message->{Body}    = 'This alias has expired.';
        return;
    }
    return 1;
}

sub add_to_alias
{
    my ($self, $alias, $id, @addresses) = @_;
    my @added = $alias->add( @addresses ) or return;
    $self->notify( $alias, $id, @added );
}

1;
__DATA__

=head1 NAME

Mail::SimpleList - module for managing simple, temporary, easy mailing lists

=head1 SYNOPSIS

    use Mail::SimpleList;
    my $list = Mail::SimpleList->new( 'alias' );
    $list->process();

=head1 DESCRIPTION

Sometimes, you just need a really simple mailing list to last for a few days.
You want it to be easy to create and easy to use, and you want it to be easy to
unsubscribe.  It's not worth setting up a complex system with one of the
existing mailing list managers, but it's nice not to worry about who does and
doesn't want to participate.

Mail::SimpleList, Mail::SimpleList::Aliases, and Mail::SimpleList::Alias make
it easy to create a temporary mailing list system.

=head1 USING LISTS

=head2 INSTALLING

Please see the README file in this distribution for installation and
configuration instructions.  You'll need to configure your mail server just a
little bit, but you only need to do it once.  The rest of these instructions
assume you've installed Mail::SimpleList to use the address
C<alias@example.com>.

=head2 CREATING A LIST

To create a list, send an e-mail to the address C<alias@example.com>.  In the
subject of the message, include the phrase C<*new*>.  In the body of the
message, include a list of e-mail addresses to be subscribed to the list.  For
example, you may create a list including Alice, Bob, and Charlie with an e-mail resembling:

    From:    you@example.com
    To:      alias@example.com
    Subject: *new*

    alice@example.com
    bob@example.com
    charlie@example.com

You will receive a response informing you that the list has been created.
Alice, Bob, and Charlie will each receive a response indicating that you have
subscribed them to the alias.  Each message will include the alias-specific
posting address.  In this case, it might be C<alias+3abfeec@example.com>.

You can specify additional directives when creating a list.  Please see
L</DIRECTIVES> for more information.

=head2 POSTING TO A LIST

To post to a list, simply send an e-mail to the address provided in the
subscription or creation message.  It will be sent to everyone currently
subscribed to the list.  You do not need to use the Reply-All feature on your
mailer; Mail::SimpleList sets the C<Reply-To> header automatically.

If you do include other e-mail addresses in the C<CC> header, Mail::SimpleList
will, by default, automatically subscribe them to the list, informing them of
this action and sending them the current message.  No one should receive
duplicate messages, even if he is already subscribed.

=head2 UNSUBSCRIBING FROM A LIST

To unsubscribe from a list, send an e-mail to the address provided with a
subject line of C<*unsubscribe*>.  In Bob's case, this message might be:

    From:    bob@example.com
    To:      alias+3abfeec@example.com
    Subject: *unsubscribe*

    no body here; it doesn't matter

Bob will receive an e-mail confirming that he has been unsubscribed.  He will
not receive any more messages directed to this list unless he resubscribes.

=head2 CLONING A LIST

To clone a list, duplicating its subscribers but setting yourself as the owner
and changing other attributes, send a message to the main address (in this
case, C<alias@example.com>) with a subject containing the command C<*clone*>
and the address of the list to clone.  Alice could clone the list above by
sending the message:

    From:    alice@example.com
    To:      alias@example.com
    Subject: *clone* alias+3abfeec@example.com

    doug@example.com

Alice will receive a list creation message and the members of the cloned list
will each receive a message informing them that they have been added to the
clone.  Doug will also be added to this new list.

=head1 DIRECTIVES

Lists have seven attributes.  Creating a list will set two attributes, the list
owner and the list members.  You can specify the other attributes by including
directives when you create or clone a list.

Directives go in the body of the creation or clone message, B<before> the list
of e-mail addresses to add.  They take the form:

    Directive: option

=head2 Closed

This directive governs whether or not the list is closed to non-subscribers.
If so, only members of the list may send a message to the list.  Everyone else
will receive an error message indicating that the list is closed.

This attribute is false by default; anyone can post to a list.  To enable it,
use either directive form:

    Closed: yes
    Closed: 1

=head2 Expires

This directive governs how long the list will be available.  After its
expiration date has passed, no one may send a message to the list.  Everyone
will then receive an error message indicating that the list has expired.

This attribute is not set by default; lists do not expire.  To enable it, use
the directive form:

    Expires: 7d2h

This directive will cause the list to expire in seven days and two hours.
Valid time units are:

=over 4

=item * C<m>, for minute.  This is sixty (60) seconds.

=item * C<h>, for hour.  This is sixty (60) minutes.

=item * C<d>, for day.  This is twenty-four (24) hours.

=item * C<w>, for week. This is seven (7) days.

=item * C<M>, for month.  This is thirty (30) days.

=back

This should suffice for most purposes.

=head2 Auto_add

This directive governs whether addresses found in the C<CC> header will be
added automatically to the list.  By default, it is enabled.  To disable it,
use either directive form:

    Auto_add: no
    Auto_add: 0

=head2 Description

This is a single line that describes the purpose of the list.  It is sent to
everyone when they are added to the list.  By default, it is blank.  To set a
description, use the form:

    Description: A list for discussing fluoridation.

=head2 Name

This isn't a directive in the sense that it's an intrinsic part of a list.
It's just a way to give a list a nicer name when it's being created.  Instead
of having to rename a list to a nicer name, you can specify it with:

    Name: meat-eaters

Only letters, numbers, dashes, and the underscore characters are allowed in
names.

=head1 METHODS

=over 4

=item * new( $alias_directory, [ $filehandle ] )

C<new()> takes one mandatory argument and one optional argument.
C<$alias_directory> is the path to the directory where alias data is stored.
C<$filehandle> is a filehandle (or a reference to a glob) from which to read an
incoming message.  By default, it will read from C<STDIN>, as that is how mail
filters work.

=item * C<process()>

Processes one incoming message.

=item * C<expansion_pattern()>

Returns a compiled regex to find expanded e-mail addresses (of the form
C<you+expansion@example.com>).  If you've set your mail server to use a
delimiter other than C<+>, override this method.  For example, Andy Lester uses
addresses of the form C<you-expansion@example.com>.  What a nut.

=item * C<add_signature( $signature )>

Adds the given string signature to the message, using the standard marker for
signatures.  If this message is a multi-part message, creates a new plain text
attachment.  Right now it only respects ASCII signatures.

=item * C<add_to_alias( $alias, $post_address, @addresses )>

Adds each of the given C<@addresses> to the C<$alias>, notifying them of the
addition and giving them the C<$post_address> of the list (so they can post to
it).  This will only notify those addresses actually added to the list.

=item * C<can_deliver( $alias, $message_headers )>

Given an Alias object for the list and the C<$message_headers>, returns true if
the sender can actually send a message to the list.  Currently, all users can
post to any list unless the list is closed to non-member postings or has
expired.

=item * C<deliver( $alias)>

Delivers the message, first checking C<can_deliver()>, then adding all new
addresses if auto-add is in effect, then setting the appropriate headers (such
as list headers and the list description), and finally adding a signature and
delivering the message.

This is currently a lot of responsibility.  Don't rely on everything all
happening in this magic method.

=item * C<generate_alias( [ $suggested_list_id ] )>

Generates a new alias for a list, given an optional list id.  The current
algorithm uses the current time plus a random number to avoid collisions with
another list alias.  It's mostly race-condition free.

=item * C<notify( $alias, $list_address, @addresses )>

Sends a message from the list owner to each address in C<@addresses>, notifying
them that the system has added them to the C<$alias> with the posting address
of C<$list_address>.

=item * C<parse_alias( $address )>

Tries to discern the id of the alias from the given string containing an
e-mail address.  This will return the empty string if it could not find one.

=item * C<post_address( $list_id )>

Given the id of the list, returns the e-mail address to use to post to the list
in string form.  This depends on your MTA setting the proper delivery headers.

=item * C<reject( [ $reason ] )>

Rejects processing this message by dying with the error code 100.  If there is
no C<$reason>, this reports an invalid alias.

This method tends to end C<procmail> processing, making the MDA report back to
the sender, if so configured.

=back

=head1 AUTHOR

chromatic, C<chromatic at wgz dot org>, with suggestions from various friends
and family as well as the subset of the Portland Perl Mongers who do not appear
in a previous category.

=head1 BUGS

No known bugs, though something odd happens when you send a message with an
invalid command to the base address (in our example, C<alias@example.com>).

=head1 TODO

=over 4

=item * look for addable addresses in C<To> header as well

=item * allow the list owner to add people even if C<Auto_add> is off

=item * explore appropriate mailing-list and references headers

=item * explicitly forbid loops (strip out all other alias addresses when
sending)

=item * allow list creation to be restricted to a set of users

=item * show lists of which I am a member

=back

=head1 COPYRIGHT

Copyright (c) 2016 chromatic.  All rights reserved.  This module is distributed
under the same terms as Perl 5.24, in the hope that it is useful but under no
warranty.
